crates/qasm3/src/exporter.rs

```diff
@@ -0,0 +1,1412 @@
+            string: name.clone(),
+        };
+        let _ = self.bind(&name);
+        self.set_reginfo(
+            IdentifierOrSubscripted::Identifier(identifier.clone()),
+            (*register).to_owned(),
+        );
+        Ok(identifier)
+    }
+
+    fn push_scope(&mut self) {
+        self.symbols.push(HashMap::new());
+        self.bitinfo.push(HashMap::new());
+        self.reginfo.push(HashMap::new());
+    }
+
+    fn pop_scope(&mut self) {
+        self.symbols.pop();
+        self.bitinfo.pop();
+        self.reginfo.pop();
+    }
+
+    fn new_context(&mut self) -> Self {
+        let mut new_table = SymbolTable::new();
+        new_table.gates.clone_from(&self.gates);
+        new_table.stdgates.clone_from(&self.stdgates);
+        new_table
+    }
+}
+
+pub struct Exporter {
+    includes: Vec<String>,
+    basis_gates: Vec<String>,
+    disable_constants: bool,
+    allow_aliasing: bool,
+    indent: String,
+}
+
+impl Exporter {
+    pub fn new(
+        includes: Vec<String>,
+        basis_gates: Vec<String>,
+        disable_constants: bool,
+        allow_aliasing: bool,
+        indent: String,
+    ) -> Self {
+        Self {
+            includes,
+            basis_gates,
+            disable_constants,
+            allow_aliasing,
+            indent,
+        }
+    }
+
+    pub fn dumps(&self, circuit_data: &CircuitData, islayout: bool) -> ExporterResult<String> {
+        let mut builder = QASM3Builder::new(
+            circuit_data,
+            islayout,
+            self.includes.clone(),
+            self.basis_gates.clone(),
+            self.disable_constants,
+            self.allow_aliasing,
+        );
+        match builder.build_program() {
+            Ok(program) => {
+                let mut output = String::new();
+                BasicPrinter::new(&mut output, self.indent.to_string(), false)
+                    .visit(&Node::Program(&program));
+                Ok(output)
+            }
+            Err(e) => Err(QASM3ExporterError::Error(e.to_string())),
+        }
+    }
+
+    pub fn dump<W: Write>(
+        &self,
+        circuit_data: &CircuitData,
+        islayout: bool,
+        writer: &mut W,
+    ) -> ExporterResult<()> {
+        let mut builder = QASM3Builder::new(
+            circuit_data,
+            islayout,
+            self.includes.clone(),
+            self.basis_gates.clone(),
+            self.disable_constants,
+            self.allow_aliasing,
+        );
+
+        match builder.build_program() {
+            Ok(program) => {
+                let mut output = String::new();
+                let mut printer = BasicPrinter::new(&mut output, self.indent.to_string(), false);
+                printer.visit(&Node::Program(&program));
+                drop(printer);
+                let _ = writer.write_all(output.as_bytes());
+                Ok(())
+            }
+            Err(e) => Err(QASM3ExporterError::Error(e.to_string())),
+        }
+    }
```

Comment:
If we need both of these, one of them should be defined in terms of the other, most likely `dumps` in terms of `dump`.

Currently `dump` is offering no performance benefit over a user manually creating the string and calling `writer.write_all(output.as_bytes())` themselves, but the API makes it look like it does.  Imo, in Rust, it's generally more idiomatic to make expensive compound operations look expensive by making the user do them.